src/js/lib/components/smartTextInput/smartTextInput.jsx

```diff
@@ -0,0 +1,208 @@
+import React from 'react';
+import CustomTextInput from '../customTextInput';
+import SuggestionList from './subcomponents/suggestionList';
+
+class SmartTextInput extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            focused: false,
+            value: props.value,
+            selectedItemId: props.valueId,
+            hoveredItemId: null,
+            hoveredItemIndex: null,
+            showSuggestions: false
+        };
+
+        this.handleInputChange = this.handleInputChange.bind(this);
+        this.handleFocus = this.handleFocus.bind(this);
+        this.handleBlur = this.handleBlur.bind(this);
+        this.handleSuggestionSelect = this.handleSuggestionSelect.bind(this);
+        this.handleSuggestionHover = this.handleSuggestionHover.bind(this);
+        this.handleClick = this.handleClick.bind(this);
+        this.handleKeyDown = this.handleKeyDown.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        const newState = {};
+        if(JSON.stringify(nextProps.suggestionIndex) !== JSON.stringify(this.props.suggestionIndex)) {
+            newState.hoveredItemId = null;
+            newState.hoveredItemIndex = null;
+            newState.selectedItemId = null;
+        }
+
+        if(nextProps.value !== this.state.value) {
+            newState.value = nextProps.value;
+        }
+
+
+        this.setState(newState);
+    }
+    
+    handleInputChange(event) {
+        this.setState({
+            value: event.target.value,
+            showSuggestions: true,
+            hoveredItemId: null,
+            hoveredItemIndex: null
+        });
+        this.props.onChange(event.target.value);
+    }
+
+    handleFocus(event) {
+        this.setState({
+            focused: true,
+            showSuggestions: true
+        });
+
+        this.props.onFocus(event);
+    }
+
+    handleBlur(event) {
+        this.setState({
+            focused: false,
+            showSuggestions: false
+        });
+
+        this.props.onBlur(event);
+    }
+
+    handleSuggestionSelect(id) {
+        this.setState({
+            selectedItemId: id,
+            value: this.props.suggestionIndex[id],
+            showSuggestions: false
+        });
+        this.props.onSelect(id, this.props.suggestionIndex[id]);
+    }
+
+    handleSuggestionHover(id, index) {
+        this.setState({
+            hoveredItemId: id,
+            hoveredItemIndex: index
+        });
+    }
+
+    handleClick(event) {
+        if(this.state.focused) {
+            this.setState({
+                showSuggestions: true
+            });
+        }
+    }
+
+    handleKeyDown(event) {
+        // 40 - down
```

Comment:
Would be helpful to eventually refactor these values out into constants. I won't request changes, but you should make a note of that in a zube task somewhere.